### ISSUE TITLE
[PERF] Regression in PostgreSQL Active Records methods.

### ISSUE THREAD

**Issue by tgxworld** [rails/rails#22081]
Creating an issue for [commit](https://github.com/rails/rails/commit/cbcdecd2c55fca9613722779231de2d8dd67ad02#commitcomment-13951842) in case it gets lost.

https://github.com/rails/rails/commit/cbcdecd2c55fca9613722779231de2d8dd67ad02 caused a regression in some of the PostgreSQL benchmarks so I'm bringing it up for visibility.

![screenshot from 2015-10-23 21 12 15](https://cloud.githubusercontent.com/assets/4335742/10693695/51d3e522-79cc-11e5-8859-2bd8b6b875e2.png)
![screenshot from 2015-10-23 21 11 33](https://cloud.githubusercontent.com/assets/4335742/10693696/51d52fe0-79cc-11e5-8b88-b4c64e9b3d6e.png)
![screenshot from 2015-10-23 21 11 22](https://cloud.githubusercontent.com/assets/4335742/10693697/56b59248-79cc-11e5-8b86-085c54e2267a.png)

Benchmark graphs over at http://rubybench.org/rails/rails/commits

Replicated on of the regressions on my local machine.
Benchmarkscript: https://github.com/ruby-bench/ruby-bench-suite/blob/b6466b277c49d8917a0a9ea8f113fbc470407143/rails/benchmarks/bm_activerecord_scope_where.rb
```
# Before
activerecord/postgres_scope_where 946.737059083313/ips

# After
activerecord/postgres_scope_where 765.2867424368787/ips
```

Let me know if I can help in anyway :)

**Comment by sgrif** [comment]
This is likely to be expected, as I've explicitly made a trade off against that form of `where`. Do we have a benchmark showing the hash form?

**Comment by tgxworld** [comment]
Just to clarify, does hash form refer to `where("foo = ?", 1)` or `where(foo: 1)`?

**Comment by sgrif** [comment]
`where(foo: 1)`

**Comment by tgxworld** [comment]
Ah we're mixing both right now in a single [benchmark](https://github.com/ruby-bench/ruby-bench-suite/blob/b6466b277c49d8917a0a9ea8f113fbc470407143/rails/benchmarks/bm_activerecord_scope_where.rb).

I'll look into splitting it up :smile:

**Comment by sgrif** [comment]
Thanks. These will have different performance characteristics. Though
theoretically we can prepare the string form as well it will be a
nontrivial amount of work atm

**Comment by kamipo** [comment]
Fixed by 2a56b2d90d4fed8548e3a1e7a7b206454858c872?
